### PR TITLE
feat: support pipeline pause with pipeline spec change

### LINKED ISSUE #14: Pause pipelines before updating them

### PR DESCRIPTION
**Pull request by xdevxy**
Fixes #14 

### Modifications

This introduced the pipeline pausing logic when pipeline spec is updated. However, currently this PR always disable pause as https://github.com/numaproj/numaplane/issues/4 is not there yet.

### Verification
Local cluster testing

### CODE REVIEW

File: config/manager/numaflow-controller-definitions-config.yaml [hunk 1]
```diff
@@ -50,7 +50,25 @@ data:
...
+                    - name: NUMAFLOW_LEADER_ELECTION_LEASE_RENEW_DEADLINE
+                      valueFrom:
+                        configMapKeyRef:
+                          key: controller.leader.election.lease.renew.deadline
+                          name: numaflow-cmd-params-config
+                          optional: true
+                    - name: NUMAFLOW_LEADER_ELECTION_LEASE_RENEW_PERIOD
+                      valueFrom:
+                        configMapKeyRef:
+                          key: controller.leader.election.lease.renew.period
```
Comment (juliev0) [inline config/manager/numaflow-controller-definitions-config.yaml:71]: any idea why all of the stuff in this file changed?

File: internal/kubernetes/dynamic_client.go [hunk 2]
```diff
@@ -99,7 +125,7 @@ func ApplyCRSpec(ctx context.Context, restConfig *rest.Config, object *GenericOb
 		}
 
 	} else {
-		numaLogger.Debugf("found existing Resource definition for %s/%s: %+v", object.Namespace, object.Name, resource)
+		numaLogger.Debugf(fmt.Sprintf("found existing Resource definition for %s/%s: %+v", object.Namespace, object.Name, resource))
```
Comment (juliev0) [inline internal/kubernetes/dynamic_client.go:128]: why is this needed?

File: internal/controller/pipelinerollout_controller.go [hunk 3]
```diff
@@ -138,14 +147,16 @@ func (r *PipelineRolloutReconciler) reconcile(ctx context.Context, pipelineRollo
...
-		return nil
+		return false, nil
 	}
 
 	// add Finalizer so we can ensure that we take appropriate action when CRD is deleted
 	if !controllerutil.ContainsFinalizer(pipelineRollout, finalizerName) {
 		controllerutil.AddFinalizer(pipelineRollout, finalizerName)
 	}
 
+	// if not, check the change if it is image change, if yes, pause
```
Comment (juliev0) [inline internal/controller/pipelinerollout_controller.go:158]: I think this comment no longer applies, right?

File: internal/controller/pipelinerollout_controller.go [hunk 4]
```diff
@@ -161,24 +172,80 @@ func (r *PipelineRolloutReconciler) reconcile(ctx context.Context, pipelineRollo
...
+			return false, nil
+		}
+
+		// If pipeline status is not above, detect if pausing is required.
+		shouldPause, err := needsPausing(pipeline, &obj)
+		if err != nil {
+			return false, err
+		}
+		if shouldPause {
+			// Use the existing spec, then pausing and re-enqueue
```
Comment (juliev0) [inline internal/controller/pipelinerollout_controller.go:224]:
```suggestion
			// Use the existing spec, then pause and re-enqueue
```

File: config/manager/numaflow-controller-definitions-config.yaml (hunk 1 above)
Comment (xdevxy) [reply config/manager/numaflow-controller-definitions-config.yaml:71]: ah because I updated the numaflow controller to match v1.2.1.

File: internal/kubernetes/dynamic_client.go (hunk 2 above)
Comment (xdevxy) [reply internal/kubernetes/dynamic_client.go:128]: the whole thing is a string, right?

Comment (juliev0) [reply internal/kubernetes/dynamic_client.go:128]: `Debugf` should resolve the formatting by itself

Comment (xdevxy) [reply internal/kubernetes/dynamic_client.go:128]: then, you don't need  `%s/%s: %+v` IIUC.

Comment (xdevxy) [reply internal/kubernetes/dynamic_client.go]: revert this back as this is `Debugf` which should be the desired behavior. I was not able to see the complete messages which should be a logger issue then.

**Review by juliev0 — APPROVED** [review]
